### PR TITLE
DGA Detection - DLTK

### PR DESCRIPTION
**Pull request by nsreekanta**
This PR is for DGA detection for ESCU using DSDL. The detection uses a deep learning model to detect DGA domains. The model is pre-trained and is available in S3 bucket.  The model needs to be downloaded and deployed within DLTK and the deployment steps are outlined here - https://github.com/splunk/security_content/wiki/How-to-deploy-Deep-Learning-models-for-ESCU.
The detection has 'apply' command to predict domains using the deployed model. 

`| tstats `security_content_summariesonly` values(DNS.answer) as IPs min(_time) as firstTime  max(_time) as lastTime from datamodel=Network_Resolution by DNS.src, DNS.query 
| `drop_dm_object_name("DNS")` | rename query AS domain | fields IPs, src, domain, firstTime, lastTime, src | apply dga_model_dltk | rename pred_dga_proba AS dga_score | where dga_score>0.5 | table src, domain, IPs, firstTime, lastTime,dga_score`

The model outputs a dga_score that indicates how likely is the domain a DGA-generated domain. So a domain that has a dga_score greater than 0.5, (the threshold set in the detection) then the domain is classified as a DGA domain. Below is an example of the detection results. 

Screen Shot 2022-09-23 at 3 30 12 PM

### CODE REVIEW

File: detections/experimental/network/detect_dga_domains_using_dltk.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  - DNS Hijacking
+  - Suspicious DNS Traffic
+  - Dynamic DNS
+  - Command and Control
+  asset_type: Web Server
+  cis20:
+  - CIS 8
+  - CIS 12
+  - CIS 13
+  confidence: 40
```
Comment (patel-bhavin) [inline detections/experimental/network/detect_dga_domains_using_dltk.yml]: change to 90

File: detections/experimental/network/detect_dga_domains_using_dltk.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  cve:
```
Comment (patel-bhavin) [inline detections/experimental/network/detect_dga_domains_using_dltk.yml]: remove this

File: detections/experimental/network/detect_dga_domains_using_dltk.yml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  - CVE-2022-32154
+  context:
+  - Source:Endpoint
+  - Stage:Initial Access
+  dataset:
+  - https://github.com/splunk/attack_data/raw/master/datasets/attack_techniques/T1203/search_activity.txt
+  impact: 50
+  kill_chain_phases:
+  - Actions on Objectives
```
Comment (patel-bhavin) [inline detections/experimental/network/detect_dga_domains_using_dltk.yml]: remove action on objectives

File: detections/experimental/network/detect_dga_domains_using_dltk.yml [hunk 4]
```diff
@@ -0,0 +1,59 @@
+name: Detect DGA domains using Pretrained Deep Learning Model in DLTK
+id: b4aefb5f-1037-410d-a149-1e091288ba33
```
Comment (patel-bhavin) [inline detections/experimental/network/detect_dga_domains_using_dltk.yml]: create a new UUID

File: detections/experimental/network/detect_dga_domains_using_dltk.yml [hunk 5]
```diff
@@ -0,0 +1,59 @@
+name: Detect DGA domains using Pretrained Deep Learning Model in DLTK
```
Comment (patel-bhavin) [inline detections/experimental/network/detect_dga_domains_using_dltk.yml]: name of the file same as the search name

File: detections/experimental/network/detect_dga_domains_using_dltk.yml [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+description: 'The following analytic uses a pretrained deep learning model to detect Domain Generation Algorithm (DGA) generated domains. The model is trained independently and then made available in AWS S3. One of the prominent indicators of a domain being DGA generated is if the domain name consists of unusual character sequences or concatenated dictionary words. Adversaries often use clever techniques to obfuscate machine generated domain names as human generated. Predicting DGA generated domain names requires analysis and building a model based on carefully chosen features. The deep learning model we have developed uses the domain name to anlayze patterns of character sequences along with carefully chosen custom features to predict if a domain is DGA generated. The model takes a domain name consisting of second-level and top-level domain names as input and outputs a dga_score. Higher the dga_score, the more likely the input domain is a DGA domain. The threshold for flagging a domain as DGA is set at 0.5.'
+search: "| tstats `security_content_summariesonly` values(DNS.answer) as IPs min(_time) as firstTime from datamodel=Network_Resolution by DNS.src, DNS.query
```
Comment (patel-bhavin) [inline detections/experimental/network/detect_dga_domains_using_dltk.yml]: improve the search output fields in the table command.

Comment (patel-bhavin) [reply detections/experimental/network/detect_dga_domains_using_dltk.yml]: add max(_time) as lastTime

File: detections/experimental/network/detect_dga_domains_using_dltk.yml [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  - Command & Control
+  message: A potentially a DGA domain $domain$ was detected, kindly review.
+  mitre_attack_id:
+  - T1189
+  nist:
+  - PR.DS
+  - PR.PT
+  - DE.AE
+  - DE.CM
+  observable:
```
Comment (patel-bhavin) [inline detections/experimental/network/detect_dga_domains_using_dltk.yml]: eg: https://github.com/splunk/security_content/blob/develop/detections/network/plain_http_post_exfiltrated_data.yml#L43

observable:
  - name: query
    type: URL
    role:
    - Other
  - name: src
    type: Hostname
    role:
    - Victim

File: detections/experimental/network/detect_dga_domains_using_dltk.yml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  product:
+  - Splunk Enterprise
+  - Splunk Enterprise Security
+  - Splunk Cloud
+  required_fields:
+  - _time
```
Comment (patel-bhavin) [inline detections/experimental/network/detect_dga_domains_using_dltk.yml]: add all field names used in the search

File: detections/experimental/network/detect_dga_domains_using_dltk.yml [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  - Source:Endpoint
+  - Stage:Initial Access
+  dataset:
+  - https://github.com/splunk/attack_data/raw/master/datasets/attack_techniques/T1203/search_activity.txt
+  impact: 50
+  kill_chain_phases:
+  - Actions on Objectives
+  - Command & Control
+  message: A potentially a DGA domain $domain$ was detected, kindly review.
+  mitre_attack_id:
```
Comment (patel-bhavin) [inline detections/experimental/network/detect_dga_domains_using_dltk.yml]: https://attack.mitre.org/techniques/T1568/002/

T1568.002

File: detections/experimental/network/detect_dga_domains_using_dltk.yml [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+  cis20:
+  - CIS 8
+  - CIS 12
+  - CIS 13
+  confidence: 40
+  cve:
+  - CVE-2022-32154
+  context:
+  - Source:Endpoint
+  - Stage:Initial Access
```
Comment (patel-bhavin) [inline detections/experimental/network/detect_dga_domains_using_dltk.yml]: Command and Control

File: detections/experimental/network/detect_dga_domains_using_dltk.yml [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+search: "| tstats `security_content_summariesonly` values(DNS.answer) as IPs min(_time) as firstTime from datamodel=Network_Resolution by DNS.src, DNS.query 
+| `drop_dm_object_name("DNS")` | rename query AS domain | fields domain | apply dga_model_dltk | rename pred_dga_proba AS dga_score | where dga_score>0.5 | table domain, dga_score" 
+how_to_implement: This detection depends on the DSDL app which can be found here - https://splunkbase.splunk.com/app/4607/ and the Network Resolution datamodel which can be found here - https://splunkbase.splunk.com/app/1621/. The dga_score threshold is set at 0.5 and this should be adjusted based on the environment. The detection uses a pre-trained deep learning model that needs to be deployed in DLTK. Steps for deployment are outlined here - ** update gist here **.
```
Comment (patel-bhavin) [inline detections/experimental/network/detect_dga_domains_using_dltk.yml]: update How to

File: detections/experimental/network/detect_dga_domains_using_pretrained_model_in_dsdl.yml [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+| `drop_dm_object_name("DNS")` | rename query AS domain | fields IPs, src, domain, firstTime, lastTime, src | apply pretrained_dga_model_dsdl | rename pred_dga_proba AS dga_score | where dga_score>0.5 | `security_content_ctime(firstTime)`  | `security_content_ctime(lastTime)` | table src, domain, IPs, firstTime, lastTime, dga_score | `detect_dga_domains_using_pretrained_model_in_dsdl_filter`'
+how_to_implement: "
+* Steps to deploy DGA detection model into Splunk App DSDL. This detection depends on the Splunk app for Data Science and Deep Learning which can be found here - https://splunkbase.splunk.com/app/4607/ and the Network Resolution datamodel which can be found here - https://splunkbase.splunk.com/app/1621/. The detection uses a pre-trained deep learning model that needs to be deployed in DSDL app. Follow the steps for deployment here - https://github.com/splunk/security_content/wiki/How-to-deploy-Deep-Learning-models-for-ESCU. 
+* Download the artifacts .tar.gz file from the S3 bucket - https://seal.splunkresearch.com.s3-us-west-2.amazonaws.com/pretrained_dga_model_dsdl.tar.gz
```
Comment (patel-bhavin) [inline detections/experimental/network/detect_dga_domains_using_pretrained_model_in_dsdl.yml]: needs update.

Comment (nsreekanta) [reply detections/experimental/network/detect_dga_domains_using_pretrained_model_in_dsdl.yml]: Updated

**Review by patel-bhavin — APPROVED** [review]

**Pull request merged**